tsdb/engine/tsm1/compact.go

```diff
@@ -578,11 +573,15 @@ func (c *DefaultPlanner) Plan(lastWrite time.Time) ([]CompactionGroup, int64) {
 	// each generation in descending break once we see a file less than 4.
 	end := 0
 	start := 0
-	for i, g := range generations {
-		if g.level() <= 3 {
+	// Walk the generations in reverse to find the last level 4 generation
+	// The backwards iteration is to keep nested lower level generations from preventing
+	// compactions of higher level, later generations.
+	// Test: https://goplay.tools/snippet/t5qSsHiaWqW
```

Comment:
If this is something that needs testing, it should be broken into a function / method that can have an automated test written for it.

Reply:
The test was just to make sure I had no off-by-one errors in the loop.  I'll remove the comment line.